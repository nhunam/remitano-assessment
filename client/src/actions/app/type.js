import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  CLEAR_MESSAGE: undefined
});
