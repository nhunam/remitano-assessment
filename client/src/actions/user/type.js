import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  USER_CREATE: undefined,
  USER_LOGIN: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined
});
