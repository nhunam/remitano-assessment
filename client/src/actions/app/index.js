import { ActionTypes } from "./type";
import { createActions } from "redux-actions";
//import proxy from 'http-proxy-middleware';

export const { clearMessage } = createActions({
  [ActionTypes.CLEAR_MESSAGE]: param => ({ ...param })
});
/*
export default function(app) {
  app.use(proxy('/users/**', { target: 'https://obscure-beach-46869.herokuapp.com/' }));
  app.use(proxy('/videos/**', { target: 'https://obscure-beach-46869.herokuapp.com/' }));
  app.use(proxy('/token/**', { target: 'https://obscure-beach-46869.herokuapp.com/' }));
};
*/