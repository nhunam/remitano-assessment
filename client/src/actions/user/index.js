import { ActionTypes } from "./type";
import { createActions } from "redux-actions";

export const {
  userLogin: login,
  userLogout: logout,
  userCreate: create
} = createActions({
  [ActionTypes.USER_LOGIN]: param => ({ ...param }),
  [ActionTypes.USER_CREATE]: param => ({ ...param }),
  [ActionTypes.USER_LOGOUT]: () => ({})
});
