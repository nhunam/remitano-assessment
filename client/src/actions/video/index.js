import { ActionTypes } from "./type";
import { createActions } from "redux-actions";

export const { videoFetch: fetch, videoSharing: shareVideo } = createActions({
  [ActionTypes.VIDEO_FETCH]: param => ({ ...param }),
  [ActionTypes.VIDEO_SHARING]: param => ({ ...param })
});
