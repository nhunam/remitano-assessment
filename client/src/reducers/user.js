import { ActionTypes } from "actions/user/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const userState = {
  isAuthenticated: false,
  info: {
    id: -1,
    email: "",
    token: ""
  }
};

export default {
  user: handleActions(
    {
      [ActionTypes.USER_LOGIN_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          isAuthenticated: { $set: true },
          info: {
            id: { $set: payload.id },
            email: { $set: payload.email },
            token: { $set: payload.token }
          }
        }),
      [ActionTypes.USER_LOGOUT]: state =>
        immutable(state, {
          status: { $set: AppStatus.RUNNING }
        }),
      [ActionTypes.USER_LOGOUT_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: false },
          info: { id: { $set: -1 }, email: { $set: "" }, token: { $set: "" } },
          status: { $set: AppStatus.IDLE }
        })
    },
    userState
  )
};
