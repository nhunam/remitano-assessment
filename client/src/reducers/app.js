import { ActionTypes as ActionAppTypes } from "actions/app/type";
import { ActionTypes as ActionUserTypes } from "actions/user/type";
import { ActionTypes as ActionVideoTypes } from "actions/video/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const appState = {
  status: AppStatus.IDLE,
  message: {
    type: "",
    content: ""
  }
};

export default {
  app: handleActions(
    {
      [ActionUserTypes.USER_LOGIN]: state =>
        immutable(state, {
          status: { $set: AppStatus.RUNNING }
        }),
      [ActionAppTypes.CLEAR_MESSAGE]: state =>
        immutable(state, {
          status: { $set: AppStatus.RUNNING },
          message: {
            type: { $set: "" },
            content: { $set: "" }
          }
        }),
      [ActionUserTypes.USER_LOGIN_FAILURE]: state =>
        immutable(state, {
          status: { $set: AppStatus.READY },
          message: {
            type: { $set: "danger" },
            content: { $set: "Login failed!" }
          }
        }),
      [ActionUserTypes.USER_LOGIN_SUCCESS]: state =>
        immutable(state, {
          status: { $set: AppStatus.READY },
          message: {
            type: { $set: "" },
            content: { $set: "" }
          }
        }),
      [ActionVideoTypes.VIDEO_FETCH]: (state, { payload }) =>
        immutable(state, {
          message: {
            type: { $set: "" },
            content: { $set: "" }
          }
        }),
      [ActionVideoTypes.VIDEO_SHARING]: state =>
        immutable(state, {
          status: { $set: AppStatus.RUNNING }
        }),
      [ActionVideoTypes.VIDEO_SHARING_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          message: {
            type: { $set: "success" },
            content: { $set: "Your video has been shared!" }
          },
          status: { $set: AppStatus.READY }
        }),
      [ActionVideoTypes.VIDEO_SHARING_FAILURE]: (state, { payload }) =>
        immutable(state, {
          message: {
            type: { $set: "danger" },
            content: { $set: "There are some problem has been occured!" }
          },
          status: { $set: AppStatus.READY }
        })
    },
    appState
  )
};
