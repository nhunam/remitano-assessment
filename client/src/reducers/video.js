import { ActionTypes } from "actions/video/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const videoState = {
  data: [],
  page: {
    total_pages: 1,
    has_next: false,
    has_previous: false,
    current_page: 0,
    total_elements: 0
  }
};

export default {
  videos: handleActions(
    {
      [ActionTypes.VIDEO_FETCH]: (state, { payload }) =>
        immutable(state, {
          data: { $set: payload.page_index === 0 ? [] : state.data },
          page: {
            $set:
              payload.page_index === 0
                ? {
                    total_pages: 1,
                    has_next: false,
                    has_previous: false,
                    current_page: 0,
                    total_elements: 0
                  }
                : state.page
          }
        }),
      [ActionTypes.VIDEO_FETCH_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          data: { $set: [...state.data, ...payload.data] },
          page: { $set: payload.page }
        }),
      [ActionTypes.VIDEO_FETCH_FAILURE]: (state, { payload }) =>
        immutable(state, {
          status: { $set: AppStatus.READY }
        })
    },
    videoState
  )
};
