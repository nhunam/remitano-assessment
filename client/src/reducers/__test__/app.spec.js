import { ActionTypes as ActionAppTypes } from "actions/app/type";
import { ActionTypes as ActionUserTypes } from "actions/user/type";
import { ActionTypes as ActionVideoTypes } from "actions/video/type";
import reducer from "reducers";

describe("App Reducers", () => {
  it("should return the initial state", () => {
    expect(reducer.user(undefined, {})).toMatchSnapshot();
  });

  it(`should handle ${ActionUserTypes.USER_LOGIN}`, () => {
    expect(
      reducer.app(undefined, { type: ActionUserTypes.USER_LOGIN })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionAppTypes.CLEAR_MESSAGE}`, () => {
    expect(
      reducer.app(undefined, { type: ActionAppTypes.CLEAR_MESSAGE })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionUserTypes.USER_LOGIN_FAILURE}`, () => {
    expect(
      reducer.app(undefined, { type: ActionUserTypes.USER_LOGIN_FAILURE })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionUserTypes.USER_LOGIN_SUCCESS}`, () => {
    const actions = {
      type: ActionUserTypes.USER_LOGIN_SUCCESS,
      payload: {
        data: {
          id: 1,
          name: "test",
          permissions: []
        }
      }
    };
    expect(reducer.app(undefined, actions)).toMatchSnapshot();
  });

  it(`should handle ${ActionVideoTypes.VIDEO_FETCH}`, () => {
    expect(
      reducer.app(undefined, { type: ActionVideoTypes.VIDEO_FETCH })
    ).toMatchSnapshot();
  });
  it(`should handle ${ActionVideoTypes.VIDEO_SHARING}`, () => {
    expect(
      reducer.app(undefined, { type: ActionVideoTypes.VIDEO_SHARING })
    ).toMatchSnapshot();
  });
  it(`should handle ${ActionVideoTypes.VIDEO_SHARING_SUCCESS}`, () => {
    expect(
      reducer.app(undefined, { type: ActionVideoTypes.VIDEO_SHARING_SUCCESS })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionVideoTypes.VIDEO_SHARING_FAILURE}`, () => {
    expect(
      reducer.app(undefined, { type: ActionVideoTypes.VIDEO_SHARING_FAILURE })
    ).toMatchSnapshot();
  });
});
