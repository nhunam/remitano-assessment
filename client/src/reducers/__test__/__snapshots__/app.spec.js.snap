// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Reducers should handle CLEAR_MESSAGE 1`] = `
Object {
  "message": Object {
    "content": "",
    "type": "",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle USER_LOGIN 1`] = `
Object {
  "message": Object {
    "content": "",
    "type": "",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle USER_LOGIN_FAILURE 1`] = `
Object {
  "message": Object {
    "content": "Login failed!",
    "type": "danger",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle USER_LOGIN_SUCCESS 1`] = `
Object {
  "message": Object {
    "content": "",
    "type": "",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle VIDEO_FETCH 1`] = `
Object {
  "message": Object {
    "content": "",
    "type": "",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle VIDEO_SHARING 1`] = `
Object {
  "message": Object {
    "content": "",
    "type": "",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle VIDEO_SHARING_FAILURE 1`] = `
Object {
  "message": Object {
    "content": "There are some problem has been occured!",
    "type": "danger",
  },
  "status": undefined,
}
`;

exports[`App Reducers should handle VIDEO_SHARING_SUCCESS 1`] = `
Object {
  "message": Object {
    "content": "Your video has been shared!",
    "type": "success",
  },
  "status": undefined,
}
`;

exports[`App Reducers should return the initial state 1`] = `
Object {
  "info": Object {
    "email": "",
    "id": -1,
    "token": "",
  },
  "isAuthenticated": false,
}
`;
